internal/execute/tableutil/builder_cache.go

```diff
@@ -0,0 +1,169 @@
+package tableutil
+
+import (
+	"reflect"
+
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/execute"
+	"github.com/influxdata/flux/internal/errors"
+	"github.com/influxdata/flux/plan"
+)
+
+// Constructor is the minimum interface for constructing a Table.
+type Constructor interface {
+	// Table will construct a Table from the existing contents.
+	// Invoking this method should reset the builder and all allocated
+	// memory will be owned by the returned flux.Table.
+	Table() (flux.Table, error)
+
+	// Release will release the buffered contents from the builder.
+	// This method is unnecessary if Table is called.
+	Release()
+}
+
+type BuilderContext struct {
+	Builder Constructor
+	Trigger execute.Trigger
+}
+
+// BuilderCache is an alternative table builder cache.
+// It allows a pluggable implementation of the builder.
+// The only requirement for the builder is that it must
+// implement the above Constructor interface.
+type BuilderCache struct {
+	New func(key flux.GroupKey) Constructor
+
+	tables      *execute.GroupLookup
+	triggerSpec plan.TriggerSpec
+}
+
+// Get retrieves the TableBuilder for this group key.
+// If one doesn't exist, it will invoke the New function and
+// store it within the TableBuilder.
+// If the builder was newly created, this method returns true
+// for the second parameter.
+// The interface must be a pointer to the type that is created
+// from the New method. This method will use reflection to set
+// the value of the pointer.
+func (d *BuilderCache) Get(key flux.GroupKey, b interface{}) bool {
+	ts, ok := d.lookupState(key)
+	if !ok {
+		if d.tables == nil {
+			d.tables = execute.NewGroupLookup()
+		}
+		builder := d.New(key)
+		triggerSpec := d.triggerSpec
+		if triggerSpec == nil {
+			triggerSpec = plan.DefaultTriggerSpec
+		}
+		t := execute.NewTriggerFromSpec(triggerSpec)
+		ts = BuilderContext{
+			Builder: builder,
+			Trigger: t,
+		}
+		d.tables.Set(key, ts)
+	}
+	r := reflect.ValueOf(b)
+	r.Elem().Set(reflect.ValueOf(ts.Builder))
+	return !ok
+}
+
+// Table will remove a builder from the cache and construct a flux.Table
+// from the buffered contents.
+func (d *BuilderCache) Table(key flux.GroupKey) (flux.Table, error) {
+	if d.tables == nil {
+		return nil, errors.Newf(codes.Internal, "table not found with key %v", key)
+	}
+	ts, ok := d.tables.Delete(key)
+	if !ok {
+		return nil, errors.Newf(codes.Internal, "table not found with key %v", key)
+	}
+	builder := ts.(BuilderContext).Builder
+	return builder.Table()
+}
```

Comment:
Could `lookupState` be reused here?

Reply:
I kind of wanted to, but this implementation of the cache will remove the entry after it is done so I have to use delete. The old cache doesn't do that. But, the old cache is also problematic because arrow builders conform to the implementation that's in this one. When we created the older cache, we didn't know that.